pgx/src/datum/interval.rs

```diff
@@ -0,0 +1,145 @@
+/*
+Portions Copyright 2019-2021 ZomboDB, LLC.
+Portions Copyright 2021-2022 Technology Concepts & Design, Inc. <support@tcdi.com>
+
+All rights reserved.
+
+Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+*/
+
+use std::ops::{Mul, Sub};
+
+use crate::{direct_function_call, pg_sys, FromDatum, IntoDatum, USECS_PER_SEC};
+use pg_sys::{DAYS_PER_MONTH, SECS_PER_DAY};
+use time::Duration;
+
+const MONTH_DURATION: Duration = Duration::days(DAYS_PER_MONTH as i64);
+
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct Interval(pg_sys::Interval);
+
+impl Interval {
+    pub fn from_months_days_usecs(months: i32, days: i32, usecs: i64) -> Self {
+        Interval(pg_sys::Interval {
+            day: days,
+            month: months,
+            time: usecs,
+        })
+    }
+
+    pub fn months(&self) -> i32 {
+        self.0.month
+    }
+
+    pub fn days(&self) -> i32 {
+        self.0.day
+    }
+
+    pub fn usecs(&self) -> i64 {
+        self.0.time
+    }
+}
+
+impl TryFrom<pg_sys::Datum> for Interval {
+    type Error = &'static str;
+    fn try_from(d: pg_sys::Datum) -> Result<Self, Self::Error> {
+        Ok(Interval(unsafe { *d.ptr_cast() }))
+    }
+}
+
+impl IntoDatum for Interval {
+    fn into_datum(mut self) -> Option<pg_sys::Datum> {
+        let interval = &mut self.0;
+        Some(pg_sys::Datum::from(interval as *mut _))
+    }
+    fn type_oid() -> pg_sys::Oid {
+        pg_sys::INTERVALOID
+    }
+}
+
+impl FromDatum for Interval {
+    unsafe fn from_datum(datum: pg_sys::Datum, is_null: bool) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        if is_null {
+            None
+        } else {
+            Some(datum.try_into().expect("Error converting interval datum"))
+        }
+    }
+}
+
+impl TryFrom<Duration> for Interval {
+    type Error = &'static str;
+    fn try_from(duration: Duration) -> Result<Interval, Self::Error> {
+        const INNER_RANGE_BEGIN: Duration =
+            Duration::new(-178_000_000i64 * pg_sys::SECS_PER_YEAR as i64, 0);
```

Comment:
Although the documentation specifies "178 million years" as the limit, the actual limit is somewhat higher:
```
select '178956970 years 7 months'::interval;
        interval        
------------------------
 178956970 years 7 mons
(1 row)
```
The actual limit is `i32::MAX` days. [`timestamp.h` line 47](https://doxygen.postgresql.org/datatype_2timestamp_8h_source.html#l00047) defines the relevant structure. This may make defining this conversion easier.